BZ1976328: Document the CRI-O can logs the image pull source in the info level log (#35283)
openshift/openshift-docs: modules/viewing-the-image-pull-source.adoc

// Module included in the following assemblies:
//
// *installing/validating-an-installation.adoc

[id="viewing-the-image-pull-source_{context}"]
= Viewing the image pull source

You can view the source of your pulled images by using the CRI-O logs on your nodes. 

.Prerequisites

* You have access to the cluster as a user with the `cluster-admin` role.

.Procedure

> * Review the CRI-O logs on a node or for the worker nodes:

Reviewer: It's better to use "Review the CRI-O logs on a master node or for the worker nodes", for your example use a master node. But it's a tiny problem, you can decide modify it or not.

+
[source,terminal]
----
$  oc adm node-logs ci-ln-jb5m1vb-f76d1-cfzck-master-0 -u crio
----
+
.Example output
+
The `Trying to access` log entry indicates where the images are being pulled from.
+
[source,terminal]
----
...
Aug 05 10:33:21.594996 ip-10-0-138-140.ec2.internal crio[1366]: time="2021-08-05 10:33:21.594930907Z" level=info msg="Pulling image: quay.io/openshift-release-dev/ocp-release:4.8.4-ppc64le" id=abcd713b-d0e1-4844-ac1c-474c5b60c07c name=/runtime.v1alpha2.ImageService/PullImage

> Aug 05 10:33:21.597931 ip-10-0-138-140.ec2.internal crio[1366]: time="2021-08-05 10:33:21.597839071Z" level=info msg="Trying to access \"quay.io/openshift-release-dev/ocp-release:4.8.4-ppc64le\""

Reviewer 2: It might work better to provide a more complete example here, similar to https://github.com/openshift/openshift-docs/pull/35283/files#r687011124 .

Most importantly include more than one “Trying to access” line, and explain that they were tried in that order, and the last one is the one that actually succeeded (if the pull did succeed, that is).

Reviewer 2: Ah, that’s a good point, I forgot about that: CRI-O initializes access to the same image twice, once at https://github.com/cri-o/cri-o/blob/c53e78f45ffde7a3b7a3d93d9c8c9410ca0d4fec/server/image_pull.go#L128 and once at https://github.com/cri-o/cri-o/blob/c53e78f45ffde7a3b7a3d93d9c8c9410ca0d4fec/server/image_pull.go#L229 . (FYI: @QiWang19 ). So a typical pull is indeed going to log everything twice. (If there is a _request_ for a pull, but the image already exists locally, there is one set of logs, followed by a debug-level https://github.com/cri-o/cri-o/blob/c53e78f45ffde7a3b7a3d93d9c8c9410ca0d4fec/server/image_pull.go#L157 .)

---

What I meant by “they were tried in that order, and the last one is the one that actually succeeded” is that _if_ there are multiple mirrors that are tried and fail, e.g. `quay.io/vendor/product` is configured via ICSP to be mirrored at `mirror1.local/vendor-product` and `mirror2.local/vendor-product`, but not actually available there (not copied there, or some network or permissions problem), the logs would look like this:
```
Trying to access "mirror1.local/vendor-product@digest" # Trying first mirror… fails, next attempt:
Trying to access "mirror1.local/vendor-product@digest" # Trying second mirror… fails, next attempt:
Trying to access "quay.io/vendor/product@digest" # Trying the original, succeeds
```

Or if the image were present on the second mirror, the logs would go:
```
Trying to access "mirror1.local/vendor-product@digest" # Trying first mirror… fails, next attempt:
Trying to access "mirror1.local/vendor-product@digest" # Trying second mirror… succeeds
```

Per the above note, with CRI-O, the whole log sequence would often be repeated twice, in order.

Author: @mtrmac Is that the exact wording in the logs? ^^

Reviewer 2: I’d like to defer to someone who actually has a CRI-O installation around, to make sure there are no further discrepancies. Maybe @QiWang19 ?

Reviewer 3: I will double-check the log wording.

Reviewer 3: Checked the wording in the logs.
`docker.io/library/alpine` is configured to be mirrored at `recovery.quay.io/quay/busybox` and `recovery.quay.io/libpod/busybox`, but does not exist in the mirrors.
```
level=info msg="Pulling image: docker.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c" id=8f7c20a9-c819-426c-9c3f-55ab389b5d1f name=/runtime.v1alpha2.ImageService/Pu>
level=info msg="Trying to access \"recovery.quay.io/quay/busybox@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"recovery.quay.io/libpod/busybox@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"docker.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"recovery.quay.io/quay/busybox@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"recovery.quay.io/libpod/busybox@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"docker.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Pulled image: docker.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c" 
```

`docker.io/library/alpine` is configured to be mirrored at `recovery.quay.io/quay/busybox` and `mirror.gcr.io/library/alpine`, only exists in the second mirror.

```
level=info msg="Pulling image: docker.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c" id=74da5b17-db01-4e6f-a7aa-d302924d2f49 name=/runtime.v1alpha2.ImageService/Pu>
level=info msg="Trying to access \"recovery.quay.io/quay/busybox@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"mirror.gcr.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"recovery.quay.io/quay/busybox@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"mirror.gcr.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Pulled image: docker.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c" 
```

Author: @QiWang19 Do I understand correctly, that the log is showing that it pulled an image that is not available in the mirrored repos? Is this a bug?
Are we safe to add the note that the user should look for the _Trying to access_ statement_ in the logs? If the first example is a bug, we can add a release note if someone creates a BZ.

Reviewer 3: @mburke5678 the first example is not a bug. If the image is not available in the mirrors, it tries the original source registry (like the quay.io/vendor/product@digest in the above comment https://github.com/openshift/openshift-docs/pull/35283#discussion_r689574193, `Trying to access "quay.io/vendor/product@digest" # Trying the original, succeeds`).  The _Trying to access_  still indicates the pull source of the image and it's safe to add the note.

...
----

+
[NOTE]
====
For disconnected installations, if configured correctly, even though the image is pulled from the mirrored location: 

* To view the source of pulled images, you must review the `Trying to access` log entry in the CRI-O logs. Other methods to view the image pull source, such as using the `crictl images` command on a node, show the full name of the image.  For example: `quay.io/project/image-name`. 

> * If your cluster uses an `ImageContentSourcePolicy` object to configure repository mirroring, the CRI-O logs report the full name of the image repository. For example: `quay.io/project/image-name`. 

Reviewer: I think this description is not correctly, If the cluster create an `ImageContentSourcePolicy` object to configure repository mirroring, the crio log should report mirror registry name of the image. 
@QiWang19  Can you double confirm?

Author: @lyman9966 @QiWang19 This statement came from @umohnani8 [comment in a related BZ](https://bugzilla.redhat.com/show_bug.cgi?id=1881694#c7) and is shown in [your example in that BZ](https://bugzilla.redhat.com/show_bug.cgi?id=1881694#c17).
Seeing `"Pulled image: quay.io/openshift-release-dev/...` would confuse me as a user. I wanted to clarify that this line is present in the output, but does not indicate where the image was actually pulled from. Please correct me if I am misunderstanding.

Reviewer 4: So if using an ICSP where you are mirroring `quay.io/myrepo` to `localreg/myrepo`, when you pull down `quay.io/myrepo/myiage` and do crictl images, the image will show as `quay.io/myrepo/myimage`, the original name will be used everywhere.
However, if you want to see where the image was actually pulled from, you can look at the cri-o logs and it should show something like `Pulled image: localreg/myrepo/myimage`.
Hopefully this makes sense, it can be confusing for users as they think their mirrors are not working when they actually are.

Author: @umohnani8 If I understand, the output in the BZ that shows `"Pulled image: quay.io/openshift-release-dev/...` is from the CRI-O logs that verifies that the image is pulled from the mirror repo: 

```
crio logs as below:
Mar 17 02:52:50 minmli0317gcp1-46m6l-m-0.c.openshift-qe.internal crio[1484]: time="2021-03-17 02:52:50.178706830Z" level=info msg="Checking image status: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:1926eae7cacb9c00f142ec98b00628970e974284b6ddaf9a6a086cb9af7a6c31" id=d71a23f1-17e0-40d0-b967-57f2e474bf0c name=/runtime.v1alpha2.ImageService/ImageStatus
Mar 17 02:52:50 minmli0317gcp1-46m6l-m-0.c.openshift-qe.internal crio[1484]: time="2021-03-17 02:52:50.179161853Z" level=info msg="Image quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:1926eae7cacb9c00f142ec98b00628970e974284b6ddaf9a6a086cb9af7a6c31 not found" id=d71a23f1-17e0-40d0-b967-57f2e474bf0c name=/runtime.v1alpha2.ImageService/ImageStatus
Mar 17 02:52:50 minmli0317gcp1-46m6l-m-0.c.openshift-qe.internal crio[1484]: time="2021-03-17 02:52:50.190753958Z" level=info msg="Pulling image: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:1926eae7cacb9c00f142ec98b00628970e974284b6ddaf9a6a086cb9af7a6c31" id=f740d165-4482-439f-aced-77c781f33731 name=/runtime.v1alpha2.ImageService/PullImage
Mar 17 02:52:50 minmli0317gcp1-46m6l-m-0.c.openshift-qe.internal crio[1484]: time="2021-03-17 02:52:50.194341109Z" level=info msg="Trying to access \"minmli0317gcp1.mirror-registry.qe.gcp.devcluster.openshift.com:5000/ocp/release@sha256:1926eae7cacb9c00f142ec98b00628970e974284b6ddaf9a6a086cb9af7a6c31\""
Mar 17 02:52:50 minmli0317gcp1-46m6l-m-0.c.openshift-qe.internal crio[1484]: time="2021-03-17 02:52:50.226788351Z" level=info msg="Trying to access \"minmli0317gcp1.mirror-registry.qe.gcp.devcluster.openshift.com:5000/ocp/release@sha256:1926eae7cacb9c00f142ec98b00628970e974284b6ddaf9a6a086cb9af7a6c31\""
Mar 17 02:52:52 minmli0317gcp1-46m6l-m-0.c.openshift-qe.internal crio[1484]: time="2021-03-17 02:52:52.936392518Z" level=info msg="Pulled image: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:1926eae7cacb9c00f142ec98b00628970e974284b6ddaf9a6a086cb9af7a6c31" id=f740d165-4482-439f-aced-77c781f33731 name=/runtime.v1alpha2.ImageService/PullImage
```

Reviewer 4: @mtrmac does this log look correct? When pulling an image from a mirror it just logs `Trying to access:...` but not the fact that it actually pulled from thee? Would there be a `failed to access` if it couldn't access the mirror location?

Reviewer 4: If it successfully accessed a mirror, can we log that it actually pulled the image from there in c/image? We moved these logs from debug level to info level so users can see where the image is pulled from as the image having the same name as the original location was causing confusion where users thought that their images weren't being pulled from the mirror.
@QiWang19 I remember you changed the log level, any idea if we can add a log of where it actually pulled the image from after the `Trying to access:...` logs?

Reviewer 2: Eventually there’s _some_ concern about too many logs, but this looks resonable, especially with the existing `DockerLogMirrorChoice` opt-in.

Author: @mtrmac @umohnani8 Should we proceed with this PR containing the note on the CRI-O logs reporting the full name of the image repository? Or hold until any code changes are made to improve the CRI-O logs?

.Procedure

* Review the CRI-O logs on a master node or for the worker nodes:
+
[source,terminal]
----
$  oc adm node-logs ci-ln-jb5m1vb-f76d1-cfzck-master-0 -u crio
----
+
.Example output
+
The `Trying to access` log entry indicates where the images are being pulled from.
+
[source,terminal]
----
...
Mar 17 02:52:50 ip-10-0-138-140.ec2.internal crio[1366]: time="2021-08-05 10:33:21.594930907Z" level=info msg="Pulling image: quay.io/openshift-release-dev/ocp-release:4.8.4-ppc64le" id=abcd713b-d0e1-4844-ac1c-474c5b60c07c name=/runtime.v1alpha2.ImageService/PullImage
Mar 17 02:52:50 ip-10-0-138-140.ec2.internal crio[1484]: time="2021-03-17 02:52:50.194341109Z" level=info msg="Trying to access \"li0317gcp1.mirror-registry.qe.gcp.devcluster.openshift.com:5000/ocp/release@sha256:1926eae7cacb9c00f142ec98b00628970e974284b6ddaf9a6a086cb9af7a6c31\""
Mar 17 02:52:50 ip-10-0-138-140.ec2.internal crio[1484]: time="2021-03-17 02:52:50.226788351Z" level=info msg="Trying to access \"li0317gcp1.mirror-registry.qe.gcp.devcluster.openshift.com:5000/ocp/release@sha256:1926eae7cacb9c00f142ec98b00628970e974284b6ddaf9a6a086cb9af7a6c31\""
...
----
+

> The log always shows two image pulls by default. If your internal container registry lists multiple mirrors, the images are pulled in the order listed in the registry, for example:

Reviewer 2: Something like
```suggestion
The log always shows two image pulls by default. If your ImageContentSourcePolicy lists multiple mirrors, the images are attempted to be pulled in the order listed in the configuration, for example:
```

(Also, there are corner cases - the same non-mirror registry can have multiple ICSP entries, and those are merged into a single mirror list; that merging attempts to preserve the relative mirror order, but that might not be always possible. I wouldn’t bother documenting this, though.)

> The log always shows two image pulls by default. If your internal container registry lists multiple mirrors, the images are pulled in the order listed in the registry, for example:

Reviewer 2: > The log always shows two image pulls by default

“always” and “by default” seem mutually exclusive to me. I think it’s closer to “may repeat the list two times” — a locally-present image only lists mirror attempts once, a missing one twice.

+
----
Trying to access "mirror1.local/vendor-product@digest" # Trying first mirror… fails, next attempt:

> Trying to access "mirror1.local/vendor-product@digest" # Trying second mirror… succeeds

Reviewer: line 40 should be : Trying to access "mirror2.local/vendor-product@digest" , because the crio is trying the second mirror.

Reviewer 2: Yes, my mistake.

(Also, to be clear, the ` # Trying `… comments are my explanations, not a part of the log.)

+
[source,terminal]
----
$  oc adm node-logs ci-ln-jb5m1vb-f76d1-cfzck-master-0 -u crio
----
+
.Example output
+
The `Trying to access` log entry indicates where the images are being pulled from.
+
[source,terminal]
----
...
Mar 17 02:52:50 ip-10-0-138-140.ec2.internal crio[1366]: time="2021-08-05 10:33:21.594930907Z" level=info msg="Pulling image: quay.io/openshift-release-dev/ocp-release:4.8.4-ppc64le" id=abcd713b-d0e1-4844-ac1c-474c5b60c07c name=/runtime.v1alpha2.ImageService/PullImage
Mar 17 02:52:50 ip-10-0-138-140.ec2.internal crio[1484]: time="2021-03-17 02:52:50.194341109Z" level=info msg="Trying to access \"li0317gcp1.mirror-registry.qe.gcp.devcluster.openshift.com:5000/ocp/release@sha256:1926eae7cacb9c00f142ec98b00628970e974284b6ddaf9a6a086cb9af7a6c31\""
Mar 17 02:52:50 ip-10-0-138-140.ec2.internal crio[1484]: time="2021-03-17 02:52:50.226788351Z" level=info msg="Trying to access \"li0317gcp1.mirror-registry.qe.gcp.devcluster.openshift.com:5000/ocp/release@sha256:1926eae7cacb9c00f142ec98b00628970e974284b6ddaf9a6a086cb9af7a6c31\""
...
----
+
[NOTE]
====
The log might show the image pull list twice, as shown in the preceding example. 
====
+

> If your `ImageContentSourcePolicy` object lists multiple mirrors, {product-title} attempts to pull the images in the order listed in the configuration, for example:

Reviewer: Can you tell me  what populates {product-title} ?

Author: @lyman9966 {product-title} is either OpenShift Conatiner Platform or OKD.

+
[source,terminal]
----
$  oc adm node-logs <node_name> -u crio
----
+
For example:
+
[source,terminal]
----
$  oc adm node-logs ci-ln-jb5m1vb-f76d1-cfzck-worker-0 -u crio
----
+
.Example output
+
The `Trying to access` log entry indicates where the image is being pulled from.
+
[source,terminal]
----
...
Mar 17 02:52:50 ip-10-0-138-140.ec2.internal crio[1366]: time="2021-08-05 10:33:21.594930907Z" level=info msg="Pulling image: quay.io/openshift-release-dev/ocp-release:4.8.4-ppc64le" id=abcd713b-d0e1-4844-ac1c-474c5b60c07c name=/runtime.v1alpha2.ImageService/PullImage
Mar 17 02:52:50 ip-10-0-138-140.ec2.internal crio[1484]: time="2021-03-17 02:52:50.194341109Z" level=info msg="Trying to access \"li0317gcp1.mirror-registry.qe.gcp.devcluster.openshift.com:5000/ocp/release@sha256:1926eae7cacb9c00f142ec98b00628970e974284b6ddaf9a6a086cb9af7a6c31\""
Mar 17 02:52:50 ip-10-0-138-140.ec2.internal crio[1484]: time="2021-03-17 02:52:50.226788351Z" level=info msg="Trying to access \"li0317gcp1.mirror-registry.qe.gcp.devcluster.openshift.com:5000/ocp/release@sha256:1926eae7cacb9c00f142ec98b00628970e974284b6ddaf9a6a086cb9af7a6c31\""
...
----
+

> [NOTE]

Reviewer 5: Admonitions seem like an escalation.  Sugg: See how it looks as just a para following the code block--with a `+` to keep it indented.  It's presence following the code block might be eye catching enough.  Clearly up to you.
